llvm/test/CodeGen/RISCV/rvv/fixed-vectors-insert.ll

```diff
@@ -550,9 +550,9 @@ define void @insertelt_c6_v8i64_0_add(ptr %x, ptr %y) {
 ; CHECK-NEXT:    vsetivli zero, 8, e64, m4, ta, ma
 ; CHECK-NEXT:    vle64.v v8, (a0)
 ; CHECK-NEXT:    li a2, 6
-; CHECK-NEXT:    vsetvli zero, zero, e64, m4, tu, ma
+; CHECK-NEXT:    vsetivli zero, 8, e64, m1, tu, ma
```

Comment:
This looks wrong.  8 probably doesn't fit in LMUL1.  This example happens to work out correctly, but this looks like a hint that maybe you got a VL value wrong somewhere?  (i.e. a VL larger than SmallestVT?)

Reply:
This slipped past me, not sure why VL got changed here. This patch doesn't work with VLs at all. Will take a look

Reply:
Here's what I think is happening: VL was already set to 8 before this patch, see the vsetvli a couple of instructions above, which is correct at zvl128b and LMUL=4 for v8i64. Presumably the existing lowering code uses VL=8 even for the vmv.s.x to avoid a vl toggle.

Previously the vmv.s.x and vle64.v had the same SEW/LMUL ratio and thus the same VLMAX, so RISCVInsertVSETVLI emitted the x0, x0 form of vsetvli when it noticed that VL didn't need to change:

https://github.com/llvm/llvm-project/blob/c8d405323e2f7c4ce6d65168ad8e79f669feece2/llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp#L878-L895

But now the SEW/LMUL ratio is different, hence why it no longer emits it that way. Should we go out of our way to drop 8 down to 1 just for the vmv.s.x?

Reply:
I ended up commenting on the semantics of AVL > VLMAX over on another thread.  See https://github.com/llvm/llvm-project/pull/67419#discussion_r1339044221

End result was that I basically convinced myself I was barking up the wrong tree here.  I'm going to take another look at the diff, and unless I spot something else, expect to go ahead and approve this.